library/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/recorder/SnapshotProducer.kt

```diff
@@ -6,13 +6,305 @@
 
 package com.datadog.android.sessionreplay.recorder
 
+import android.graphics.Typeface
+import android.graphics.drawable.ColorDrawable
+import android.graphics.drawable.Drawable
+import android.graphics.drawable.InsetDrawable
+import android.graphics.drawable.RippleDrawable
+import android.os.Build
+import android.view.Gravity
 import android.view.View
+import android.view.ViewGroup
+import android.view.ViewStub
+import android.widget.Button
+import android.widget.ImageView
+import android.widget.TextView
+import androidx.appcompat.widget.ActionBarContextView
+import androidx.appcompat.widget.Toolbar
+import com.datadog.android.sessionreplay.model.MobileSegment
+import java.util.LinkedList
+import java.util.Locale
 
-@Suppress("UNUSED_PARAMETER")
 internal class SnapshotProducer {
 
-    fun produce(rootView: View, scale: Float): Node? {
-        // TODO: RUMM-2356
-        return null
+    private val systemViewIds by lazy {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES
+            .LOLLIPOP
+        ) {
+            setOf(android.R.id.navigationBarBackground, android.R.id.statusBarBackground)
+        } else {
+            emptySet()
+        }
+    }
+
+    fun produce(rootView: View, pixelsDensity: Float): Node? {
+        return rootView.toNode(pixelsDensity)
+    }
+
+    @Suppress("ComplexMethod")
+    private fun View.toNode(pixelsDensity: Float): Node? {
+        if (isNotValid()) {
+            return null
+        }
+
+        if (isSystemNoise()) {
+            return null
+        }
+
+        if (isToolbar()) {
+            // skip adding the children and just take a screenshot of the toolbar. It is too complex
+            // to de - structure this in multiple wireframes and we cannot actually get all the
+            // details here.
+            return Node(
+                id = resolvedId(),
+                wireframes = listOf(this.asImageWireframeNode(pixelsDensity))
+            )
+        }
+
+        val childNodes = LinkedList<Node>()
+        if (this is ViewGroup && childCount > 0) {
+            for (i in 0 until childCount) {
+                val view = getChildAt(i) ?: continue
+                view.toNode(pixelsDensity)?.let {
+                    childNodes.add(it)
+                }
+            }
+        }
+
+        return toNode(childNodes, pixelsDensity)
+    }
+
+    private fun View.toNode(childNodes: LinkedList<Node>, pixelsDensity: Float) =
```

Comment:
Please avoid using block function when the implementation is this long, as it hides the return type and make things harder to read